src/bin/pg_autoctl/monitor.c

```diff
@@ -2031,6 +2115,247 @@ monitor_get_notifications(Monitor *monitor)
+ */
+bool
+monitor_wait_until_primary_applied_settings(Monitor *monitor,
+											const char *formation)
+{
+	PGconn *connection = monitor->pgsql.connection;
+	bool		applySettingsTransitionInProgress = false;
+	bool		applySettingsTransitionDone = false;
+
+	if (connection == NULL)
+	{
+ 		log_warn("Lost connection.");
+		return false;
+	}
+
+	log_info("Waiting for the settings to have been applied to "
+			 "the monitor and primary node");
+
+	while (!applySettingsTransitionDone)
+	{
+		/* Sleep until something happens on the connection. */
+		int         sock;
+		fd_set      input_mask;
+		PGnotify   *notify;
+
+		sock = PQsocket(connection);
+
+		if (sock < 0)
+			return false;	/* shouldn't happen */
+
+		FD_ZERO(&input_mask);
+		FD_SET(sock, &input_mask);
+
+		if (select(sock + 1, &input_mask, NULL, NULL, NULL) < 0)
+		{
+			log_warn("select() failed: %s\n", strerror(errno));
+			return false;
+		}
+
+		/* Now check for input */
+		PQconsumeInput(connection);
+		while ((notify = PQnotifies(connection)) != NULL)
+		{
+			StateNotification notification = { 0 };
+
+			if (strcmp(notify->relname, "state") != 0)
+			{
+				log_warn("%s: %s", notify->relname, notify->extra);
+				continue;
+			}
+
+			log_debug("received \"%s\"", notify->extra);
+
+			/* the parsing scribbles on the message, make a copy now */
+			strlcpy(notification.message, notify->extra, BUFSIZE);
+
+			/* errors are logged by parse_state_notification_message */
+			if (!parse_state_notification_message(&notification))
+			{
+				log_warn("Failed to parse notification message \"%s\"",
+						 notify->extra);
+				continue;
+			}
+
+			/* filter notifications for our own formation */
+			if (strcmp(notification.formationId, formation) != 0)
+			{
+				continue;
+			}
+
+			if (notification.reportedState == PRIMARY_STATE
+				&& notification.goalState == APPLY_SETTINGS_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 1/4: primary node %d (%s:%d) is assigned \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.goalState));
+			}
+			else if (notification.reportedState == APPLY_SETTINGS_STATE
+					 && notification.goalState == APPLY_SETTINGS_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 2/4: primary node %d (%s:%d) reported \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.reportedState));
+			}
+			else if (notification.reportedState == APPLY_SETTINGS_STATE
+					 && notification.goalState == PRIMARY_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 3/4: primary node %d (%s:%d) is assigned \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.goalState));
+			}
+			else if (applySettingsTransitionInProgress
```

Comment:
This code assume that any of steps 1,2,3 occurs before 4. If we ever loose some notifications and hit 4 before others our logic might break here.

Reply:
The risk does exist. I have implement a generous 30s timeout after which we exit the loop if we didn't get our notifications.